src/3d/symbols/qgspoint3dsymbol_p.cpp

```diff
@@ -460,6 +468,100 @@ Qt3DCore::QTransform *QgsModelPoint3DSymbolHandler::transform( QVector3D positio
+
+    bool prepare( const Qgs3DRenderContext &context, QSet<QString> &attributeNames ) override;
+    void processFeature( QgsFeature &feature, const Qgs3DRenderContext &context ) override;
+    void finalize( Qt3DCore::QEntity *parent, const Qgs3DRenderContext &context ) override;
+
+  private:
+
+    //! temporary data we will pass to the tessellator
+    struct PointData
+    {
+      QVector<QVector3D> positions;  // contains triplets of float x,y,z for each point
+    };
+
+    void makeEntity( Qt3DCore::QEntity *parent, const Qgs3DRenderContext &context, PointData &out, bool selected );
+
+    // input specific for this class
+    const QgsPoint3DSymbol &mSymbol;
+    // inputs - generic
+    QgsFeatureIds mSelectedIds;
+
+    // outputs
+    PointData outNormal;  //!< Features that are not selected
+    PointData outSelected;  //!< Features that are selected
+};
+
+bool QgsPoint3DBillboardSymbolHandler::prepare( const Qgs3DRenderContext &context, QSet<QString> &attributeNames )
+{
+  Q_UNUSED( context )
+  Q_UNUSED( attributeNames )
+  return true;
+}
+
+void QgsPoint3DBillboardSymbolHandler::processFeature( QgsFeature &feature, const Qgs3DRenderContext &context )
+{
+  PointData &out = mSelectedIds.contains( feature.id() ) ? outSelected : outNormal;
+
+  if ( feature.geometry().isNull() )
+    return;
+
+  Qgs3DUtils::extractPointPositions( feature, context.map(), mSymbol.altitudeClamping(), out.positions );
+}
+
+void QgsPoint3DBillboardSymbolHandler::finalize( Qt3DCore::QEntity *parent, const Qgs3DRenderContext &context )
+{
+  makeEntity( parent, context, outNormal, false );
+  makeEntity( parent, context, outSelected, true );
+}
+
+void QgsPoint3DBillboardSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const Qgs3DRenderContext &context, PointData &out, bool selected )
+{
+  // Billboard Geometry
+  QgsBillboardGeometry *billboardGeometry = new QgsBillboardGeometry();
+  billboardGeometry->setPoints( out.positions );
+
+  // Billboard Geometry Renderer
+  Qt3DRender::QGeometryRenderer *billboardGeometryRenderer = new Qt3DRender::QGeometryRenderer;
+  billboardGeometryRenderer->setPrimitiveType( Qt3DRender::QGeometryRenderer::Points );
+  billboardGeometryRenderer->setGeometry( billboardGeometry );
+  billboardGeometryRenderer->setVertexCount( billboardGeometry->count() );
+
+  // Billboard Material
+  QgsPoint3DBillboardMaterial *billboardMaterial = new QgsPoint3DBillboardMaterial();
+  QgsMarkerSymbol *symbol = mSymbol.billboardSymbol();
+
+  if ( symbol )
+  {
+    billboardMaterial->setTexture2DFromSymbol( symbol, context.map(), selected );
+  }
+  else
+  {
+    billboardMaterial->useDefaultSymbol( context.map(), selected );
+  }
+
+  // Billboard Transform
+  Qt3DCore::QTransform *billboardTransform = new Qt3DCore::QTransform();
+  billboardTransform->setMatrix( mSymbol.billboardTransform() );
+  billboardTransform->setTranslation( QVector3D( 0.0f, 0.0f, 0.0f ) + billboardTransform->translation() );
```

Comment:
is there any reason for this line? maybe some leftover from debugging?

Reply:
Yes, it seems so.